rfc/system/0000-exteranl-plugin-service.md

```diff
@@ -0,0 +1,270 @@
+    401 The request was unauthorized.
+    500 The request was not handled correctly due to a server error.
+    """
+    if resp.state == "success":
+       state = succeed
+    elif resp.state == "failure":
+       state = failed
+    elif resp.state == "running":
+        state = running
+    return {"job_spec": job_spec, "job_state": state, "error_code": error_code, "message": message}
+
+@app.delete("/plugins/v1/{task_type}/{version}")
+async def delete_res(resource_id: str):
+    """
+    Delete the resource
+    
+    return job_id, error core, message.
+    200 Job was deleted successfully.
+    400 The request was malformed. See JSON response for error details.
+    401 The request was unauthorized.
+    500 The request was not handled correctly due to a server error.
+    """
+    return {"job_id": job_id, "error_code": error_code, "message": message}
+
+```
+**Ideally**, we should provide a simplified interface to this in flytekit. This would mean that the flytekit plugin can simply use this interface to create a local plugin and a backendplugin.
+
+```python=
+class BackendPluginBase:
+
+    def __init__(self, task_type: str, version: str = "v1"):
+        self._task_type = task_type
+        self._version = version
+
+    @property
+    def task_type(self) -> str:
+        return self._task_type
+
+    @property
+    def version(self) -> str:
+        return self._version
+
+    @abstractmethod
+    async def create(self):
+        pass
+
+    @abstractmethod
+    async def poll(self):
+        pass
+    
+    @abstractmethod
+    async def delete(self):
+        pass
+    
+BackendPluginRegistry.register(BQPlugin())
+```
+
+### (Alternative) GRPC Backend Server
+```python=
+class BackendPluginServer(BackendPluginServiceServicer):
+    def CreateTask(self, request: plugin_system_pb2.TaskCreateRequest, context) -> plugin_system_pb2.TaskCreateResponse:
+        req = TaskCreateRequest.from_flyte_idl(request)
+        plugin = BackendPluginRegistry.get_plugin(req.template.type)
+        return plugin.create(req.inputs, req.output_prefix, req.template)
+
+    def GetTask(self, request: plugin_system_pb2.TaskGetRequest, context):
+        plugin = BackendPluginRegistry.get_plugin(request.task_type)
+        return plugin.get(job_id=request.job_id, output_prefix=request.output_prefix, prev_state=request.prev_state)
+
+    def DeleteTask(self, request: plugin_system_pb2.TaskDeleteRequest, context):
+        plugin = BackendPluginRegistry.get_plugin(request.task_type)
+        return plugin.delete(request.job_id)
+```
+# Register a New plugin
+
+
+### Run a Flytekit Backend Plugin
+The workflow code running Snowflake, Databricks should not be changed. We can add a new config `enable_backend_flytekit_plugin_system`. If it's true, the job will be handled by plugin system. If can't find the plugin in the registry, falling back to use the default web api plugin in the propeller.
```

Comment:
```suggestion
The workflow code running Snowflake, Databricks should not be changed. We can add a new config `enable_backend_flytekit_plugin_system`. If it's true, the job will be handled by plugin system. If it can't find the plugin in the registry, it falls back to use the default web api plugin in the propeller.
```